### ISSUE TITLE
Compilation failure on Alpine Linux

### ISSUE THREAD

**Issue by reine249** [mfem/mfem#2993]
MFEM fails to compile on Alpine Linux. I use Alpine Linux for my CI builds, so it isn't a major problem.  I've been able to repeat the error on alpine:latest as well as an older version (alpine:3.12). Below is the error message.

```bash
[ 16%] Building CXX object CMakeFiles/mfem.dir/linalg/constraints.cpp.o
In file included from /mfem/linalg/../fem/../mesh/mesh.hpp:25,
                 from /mfem/linalg/../fem/fespace.hpp:17,
                 from /mfem/linalg/constraints.cpp:14:
/mfem/linalg/../fem/../mesh/../general/zstr.hpp: In function 'std::string strict_fstream::strerror()':
/mfem/linalg/../fem/../mesh/../general/zstr.hpp:83:35: error: invalid conversion from 'int' to 'const char*' [-fpermissive]
   83 |    std::string tmp(p, std::strlen(p));
      |                                   ^
      |                                   |
      |                                   int
```

Below are the steps for repeating the error (I'm using ubuntu 20.04 Docker 20.10.7)

1. `docker run -it alpine:latest /bin/sh`
2. `apk update && apk upgrade && apk add build-base cmake git blas-dev lapack-dev openmpi-dev curl-dev`
3. Then I clone the repo, configure, and build MFEM. Failure occurs during the build process. 
    1. `git clone https://github.com/mfem/mfem.git`
    2. `cd mfem`
    3. `cmake -S. -B./build/`
    4. `cmake --build ./build/ --config Release`

**Comment by tzanio** [comment]
Thanks for reporting this @reine249, I was actually not aware of [Alpine Linux](https://www.alpinelinux.org/).

I believe the issue is likely due to the [following check](https://github.com/mfem/mfem/blob/master/general/zstr.hpp#L72-L74) not passing

```c++
#elif (_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600) && ! _GNU_SOURCE || \
      defined(__APPLE__) || defined(__FreeBSD__) || defined(__OpenBSD__) || \
      defined(__NetBSD__) || defined(__DragonFly__) || defined(__EMSCRIPTEN__)
```

Do you know if Alpine Linux defines a similar macro that we can check here?

**Comment by tzanio** [comment]
Maybe `__MUSL__`?

**Comment by reine249** [comment]
It looks like `__MUSL__` isn't defined.  After doing some digging in /usr/include/string.h, the only macro I could find that was defined in Alpine and not in Ubuntu was `__NEED_locale_t`.  However, I don't know how specific this macro is to Alpine.

Sorry for using such an obscure linux distribution.  Gitlab's examples for CI/CD use Alpine, and I found that Alpine was faster for my CI builds than Ubuntu.

**Comment by jandrej** [comment]
@reine249 Can you check that #2995 resolves this problem?

**Comment by reine249** [comment]
Yes, #2995 resolves the problem.  Thank you for the quick fix!

**Comment by tzanio** [comment]
Resolved in #2995
